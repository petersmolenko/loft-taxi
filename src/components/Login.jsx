import React, { useContext, useState } from "react";
import logo from "../assets/logo-white.svg";
import {
    Grid,
    Paper,
    Typography,
    Link,
    TextField,
    Button,
} from "@material-ui/core";
import PropTypes from "prop-types";
import { AuthContext } from "../AuthContext.jsx";

const Login = ({ setRoute }) => {
    const { login } = useContext(AuthContext);
    const [authLogin, setAuthLogin] = useState(null);
    const [authPassword, setAuthPassword] = useState(null);

    return (
        <>
            <Grid item xs={4} lg={3}>
                <img className="MainLogo" src={logo} alt="" />
            </Grid>
            <Grid item xs={7} lg={5}>
                <Paper elevation={0} className="AppForm">
                    <Typography
                        variant="h4"
                        component="h4"
                        className="AppForm_margin"
                    >
                        Войти
                    </Typography>
                    <Typography
                        variant="body1"
                        component="p"
                        className="AppForm_margin"
                    >
                        Новый пользователь?{" "}
                        <Link
                            href="/"
                            onClick={(e) => {
                                e.preventDefault();
                                setRoute("signup");
                            }}
                            color="primary"
                        >
                            Зарегистрируйтесь
                        </Link>
                    </Typography>

                    <form noValidate autoComplete="off">
                        <Grid container justify="center" direction="column">
                            <Grid item>
                                <TextField
                                    required
                                    id="userName"
                                    inputProps={{
                                        "data-testid": "userNameField",
                                    }}
                                    label="Имя пользователя"
                                    className="AppForm_margin"
                                    fullWidth={true}
                                    onChange={(e) => {
                                        setAuthLogin(e.target.value);
                                    }}
                                />
                            </Grid>
                            <Grid item>
                                <TextField
                                    required
                                    id="userPassword"
                                    label="Пароль"
                                    inputProps={{
                                        "data-testid": "userPasswordField",
                                    }}
                                    type="password"
                                    className="AppForm_margin"
                                    fullWidth={true}
                                    onChange={(e) => {
                                        setAuthPassword(e.target.value);
                                    }}
                                />
                            </Grid>
                            <Grid item align="right">
                                <Button
                                    variant="contained"
                                    color="primary"
                                    style={{
                                        background: "#ffc617",
                                        color: "rgba(0, 0, 0, 0.87)",
                                    }}
                                    data-testid="loginBtn"
                                    onClick={(e) => {
                                        e.preventDefault();
                                        login(authLogin, authPassword);
                                        setRoute("map");
                                    }}
                                >
                                    Войти
                                </Button>
                            </Grid>
                        </Grid>
                    </form>
                </Paper>
            </Grid>
        </>
    );
};

Login.propTypes = {
    setRoute: PropTypes.func,
};

export default Login;
