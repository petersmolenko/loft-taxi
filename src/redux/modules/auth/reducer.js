import {
    loggedOut,
    loggedIn,
    loggedInSuccess,
    loggedInFailure,
    signUp,
    clearErrors
} from "./actions";

export default (state = {}, action) => {
    switch (action.type) {
        case clearErrors.toString():
            return { ...state, error: null };
        case loggedOut.toString():
            return { ...state, isLoggedIn: false };
        case loggedIn.toString():
            return { ...state, isLoaded: true };
        case signUp.toString():
            return { ...state, isLoaded: true };
        case loggedInSuccess.toString():
            return {
                ...state,
                isLoggedIn: true,
                token: action.payload,
                isLoaded: false,
                error: null,
            };
        case loggedInFailure.toString():
            return { ...state, isLoaded: false, error: action.payload };
        default:
            return state;
    }
};
