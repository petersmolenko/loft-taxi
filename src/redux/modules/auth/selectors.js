const isLoggedIn = state => state.auth.isLoggedIn
const getToken = state => state.auth.token

export {isLoggedIn, getToken}