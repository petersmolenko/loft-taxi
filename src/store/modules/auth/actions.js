import {createAction} from 'redux-actions'
export const loggedIn = createAction('loft-taxi/auth/LOGGED_IN')
export const signUp = createAction('loft-taxi/auth/SIGN_UP')
export const clearErrors = createAction('loft-taxi/auth/CLEAR_ERRORS')
export const loggedInSuccess = createAction('loft-taxi/auth/LOGGED_IN_SUCCESS')
export const loggedInFailure = createAction('loft-taxi/auth/LOGGED_IN_FAILURE')
export const loggedOut = createAction('loft-taxi/auth/LOGGED_OUT')