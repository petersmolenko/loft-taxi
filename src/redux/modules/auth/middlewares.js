import { loggedIn, loggedInSuccess, loggedInFailure, signUp } from "./actions";

export const loginMiddleware = (store) => (next) => (action) => {
    const result = next(action);
    if (action.type === loggedIn.toString()) {

        fetch("https://loft-taxi.glitch.me/auth", {
            method: "POST",
            headers: {
                "Content-Type": "application/json;charset=utf-8",
            },
            body: JSON.stringify(action.payload),
        })
            .then((resp) => resp.json())
            .then((auth) => {
                if (auth.success === true) {
                    store.dispatch(loggedInSuccess(auth.token));
                } else {
                    store.dispatch(loggedInFailure(auth.error));
                }
            })
            .catch((error) => {
                store.dispatch(loggedInFailure("Error!!"));
            });
    }

    return result;
};

export const signupMiddleware = (store) => (next) => (action) => {
    const result = next(action);
    if (action.type === signUp.toString()) {
        console.log(action.payload);
        fetch("https://loft-taxi.glitch.me/register", {
            method: "POST",
            headers: {
                "Content-Type": "application/json;charset=utf-8",
            },
            body: JSON.stringify(action.payload),
        })
            .then((resp) => resp.json())
            .then((auth) => {
                if (auth.success === true) {
                    store.dispatch(loggedInSuccess(auth.token));
                } else {
                    store.dispatch(loggedInFailure(auth.error));
                }
            })
            .catch((error) => {
                store.dispatch(loggedInFailure("Error!!"));
            });
    }

    return result;
};
