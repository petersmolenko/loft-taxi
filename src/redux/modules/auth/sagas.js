import { call, put, takeLatest } from "redux-saga/effects";
import { loggedIn, loggedInSuccess, loggedInFailure, signUp } from "./actions";
import { fetchProfile } from "../profile/actions";
import {fetchAddressList} from '../routes/actions';
import {auth as apiAuth, register as apiRegister} from '../../../shared/api'

export function* authorizationSaga(action) {
    try {
        const auth = yield call(apiAuth, action.payload);

        if (auth.success === true) {
            yield put(loggedInSuccess(auth.token));
            yield put(fetchProfile());
            yield put(fetchAddressList());
        } else {
            yield put(loggedInFailure(auth.error));
        }
    } catch (e) {
        yield put(loggedInFailure(e.message));
    }
}

export function* watchAuthorizationSaga() {
    yield takeLatest(loggedIn.toString(), authorizationSaga);
}

export function* registrationSaga(action) {
    try {
        const auth = yield call(apiRegister, action.payload);
        
        if (auth.success === true) {
            yield put(loggedInSuccess(auth.token));
            yield put(fetchAddressList());
        } else {
            yield put(loggedInFailure(auth.error));
        }
    } catch (e) {
        yield put(loggedInFailure(e.message));
    }
}

export function* watchRegistrationSaga() {
    yield takeLatest(signUp.toString(), registrationSaga);
}
