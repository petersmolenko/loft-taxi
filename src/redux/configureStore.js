import { createStore, applyMiddleware, combineReducers } from 'redux';
import auth, {loginMiddleware, signupMiddleware} from './modules/auth'
import profile, {fetchProfileMiddleware} from './modules/profile'

const createStoreWithMiddleware = applyMiddleware( loginMiddleware, signupMiddleware, fetchProfileMiddleware)(createStore); 

const reducer = combineReducers({
  auth, profile
});

const configureStore = (initialState) => createStoreWithMiddleware(reducer, initialState);
export default configureStore;