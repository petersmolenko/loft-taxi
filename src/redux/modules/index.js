import { combineReducers } from "redux";
import { fork } from "redux-saga/effects";
import auth from "./auth";
import profile from "./profile";
import routes from "./routes";
import { watchFetchProfileSaga, watchPutProfileSaga } from "./profile";
import {
    watchRouteSaga,
    watchAddressListSaga,
} from "./routes";
import {
    watchAuthorizationSaga,
    watchRegistrationSaga,
} from "./auth";

export default combineReducers({ auth, profile, routes });

export function* rootSaga() {
    yield fork(watchFetchProfileSaga);
    yield fork(watchPutProfileSaga);
    yield fork(watchRouteSaga);
    yield fork(watchAddressListSaga);
    yield fork(watchAuthorizationSaga);
    yield fork(watchRegistrationSaga);
}
