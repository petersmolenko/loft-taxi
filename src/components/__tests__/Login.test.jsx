import React from "react";
import { fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Login from "../Login";
import { Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { createBrowserHistory } from "history";
import { clearErrors, loggedIn } from "../../store/modules/auth";

const mockStore = configureStore([]);

global.MutationObserver = jest.fn(function MutationObserver(callback) {
    this.observe = jest.fn();
    this.disconnect = jest.fn();
});

describe("Login Component", () => {
    it("Происходит корректная отрисовка", () => {
        const store = mockStore({
            auth: { isLoggedIn: false },
        });
        const history = createBrowserHistory();
        const tree = renderer
            .create(
                <Provider store={store}>
                    <Router history={history}>
                        <Login />
                    </Router>
                </Provider>
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("Переход по ссылке Регистрация выполняется корректно", () => {
        const store = mockStore({
            auth: { isLoggedIn: false },
        });
        const history = createBrowserHistory();
        const { getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Login />
                </Router>
            </Provider>
        );

        expect(getByText("Новый пользователь?")).toBeInTheDocument();

        fireEvent.click(getByText("Зарегистрируйтесь"));
        expect(history.location.pathname).toBe("/signup");
    });

    it("Нажатие кнопки Вход приводит к отправке валидной формы", async () => {
        const store = mockStore({
            auth: { isLoggedIn: false },
        });
        store.dispatch = jest.fn();
        const history = createBrowserHistory();
        const { getByTestId } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Login />
                </Router>
            </Provider>
        );

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(clearErrors());

        fireEvent.input(getByTestId("userNameField"), {
            target: { value: "email@email.com" },
        });
        fireEvent.input(getByTestId("userPasswordField"), {
            target: { value: "password" },
        });
        await act(async () => {
            fireEvent.click(getByTestId("loginBtn"));
        });
        expect(store.dispatch).toHaveBeenCalledTimes(2);
        expect(store.dispatch).toHaveBeenCalledWith(
            loggedIn({ email: "email@email.com", password: "password" })
        );
    });

    it("Имеется возможность изменения полей", () => {
        const store = mockStore({
            auth: { isLoggedIn: false },
        });
        const history = createBrowserHistory();
        const { getByTestId } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Login />
                </Router>
            </Provider>
        );

        fireEvent.input(getByTestId("userNameField"), {
            target: { value: "name" },
        });
        expect(getByTestId("userNameField").value).toBe("name");

        fireEvent.input(getByTestId("userPasswordField"), {
            target: { value: "password" },
        });
        expect(getByTestId("userPasswordField").value).toBe("password");
    });

    it("При нажатии кнопки Войти невалидная форма не отправляется, выводятся сообщения об ошибках", async () => {
        const store = mockStore({
            auth: { isLoggedIn: false },
        });
        store.dispatch = jest.fn();
        const history = createBrowserHistory();
        const { getByTestId, getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Login />
                </Router>
            </Provider>
        );

        fireEvent.input(getByTestId("userNameField"), {
            target: { value: "email@" },
        });
        fireEvent.input(getByTestId("userPasswordField"), {
            target: { value: "" },
        });
        await act(async () => {
            fireEvent.click(getByTestId("loginBtn"));
        });
        expect(
            getByText(`Значение поля должно иметь вид "text@text.text"!`)
        ).toBeInTheDocument();
        expect(getByText(`Поле должно быть заполено!`)).toBeInTheDocument();
    });
});
