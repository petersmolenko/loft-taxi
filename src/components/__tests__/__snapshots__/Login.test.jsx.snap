// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login testing renders correctly 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-lg-3"
  >
    <img
      alt=""
      className="MainLogo"
      src="logo-white.svg"
    />
  </div>,
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 MuiGrid-grid-lg-5"
  >
    <div
      className="MuiPaper-root AppForm MuiPaper-elevation0 MuiPaper-rounded"
    >
      <h4
        className="MuiTypography-root AppForm_margin MuiTypography-h4"
      >
        Войти
      </h4>
      <p
        className="MuiTypography-root AppForm_margin MuiTypography-body1"
      >
        Новый пользователь?
         
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
          href="/"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
        >
          Зарегистрируйтесь
        </a>
      </p>
      <form
        autoComplete="off"
        noValidate={true}
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-center"
        >
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <div
              className="MuiFormControl-root MuiTextField-root AppForm_margin MuiFormControl-fullWidth"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                data-shrink={false}
                htmlFor="userName"
                id="userName-label"
              >
                Имя пользователя
                <span
                  className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiInput-input"
                  data-testid="userNameField"
                  disabled={false}
                  id="userName"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={true}
                  type="text"
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <div
              className="MuiFormControl-root MuiTextField-root AppForm_margin MuiFormControl-fullWidth"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                data-shrink={false}
                htmlFor="userPassword"
                id="userPassword-label"
              >
                Пароль
                <span
                  className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiInput-input"
                  data-testid="userPasswordField"
                  disabled={false}
                  id="userPassword"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={true}
                  type="password"
                />
              </div>
            </div>
          </div>
          <div
            align="right"
            className="MuiGrid-root MuiGrid-item"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              data-testid="loginBtn"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "background": "#ffc617",
                  "color": "rgba(0, 0, 0, 0.87)",
                }
              }
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Войти
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
]
`;
